### PR TITLE
make use of range requests if available in werkzeug

### PR DESCRIPTION
**Pull request by magne4000**
This change is [Reviewable](https://reviewable.io/reviews/pallets/flask/2031)

### CODE REVIEW

File: flask/helpers.py [hunk 1]
```diff
@@ -575,12 +585,21 @@ def send_file(filename_or_fp, mimetype=None, as_attachment=False,
             warn('Access %s failed, maybe it does not exist, so ignore etags in '
                  'headers' % filename, stacklevel=2)
 
-        if conditional:
+    if conditional:
+        try:
+            # Added in werkzeug 0.12
+            rv = rv.make_conditional(request, accept_ranges=True,
+                                     complete_length=fsize)
+        except TypeError:
```
Comment (untitaker) [inline flask/helpers.py]: This seems a bit broad. Could we change this to a version check for Werkzeug, or check for existence of a function like done in the tests?

File: flask/helpers.py [hunk 2]
```diff
@@ -575,12 +585,21 @@ def send_file(filename_or_fp, mimetype=None, as_attachment=False,
...
             rv = rv.make_conditional(request)
-            # make sure we don't send x-sendfile for servers that
-            # ignore the 304 status code for x-sendfile.
-            if rv.status_code == 304:
-                rv.headers.pop('x-sendfile', None)
+        except RequestedRangeNotSatisfiable:
+            if close_on_416:
```
Comment (untitaker) [inline flask/helpers.py]: I don't understand why this is necessary, why not always close?

Comment (RonnyPfannschmidt) [inline flask/helpers.py]: as far as i can tell this close specifically closes only if the file was opened in order to maintain unchanged outside ressources

perhaps a more speaking variable name would help

Comment (untitaker) [inline flask/helpers.py]: Hmm, but the file is always closed if one uses `x-sendfile`. So I suspect it's expected that the file is always closed by `send_file`.

File: flask/helpers.py (hunk 1 above)
Comment (magne4000) [inline flask/helpers.py]: I'm not found of testing a version for functionnalities, I'd rather test like you suggested for the existence of a function. But I find this less "obvious". At least this test is clear, even if a bit broad.

File: flask/helpers.py (hunk 2 above)
Comment (magne4000) [inline flask/helpers.py]: @RonnyPfannschmidt is right, that was why I did it that way.
Do we decide that we want to always close (if closable) the file ?

File: flask/helpers.py (hunk 1 above)
Comment (untitaker) [inline flask/helpers.py]: But the less obvious code is more correct, and I think that trumps most other aspects.

File: flask/helpers.py (hunk 2 above)
Comment (untitaker) [inline flask/helpers.py]: I think the file is not closed in some cases because otherwise it wouldn't be possible to stream it.

Comment (magne4000) [inline flask/helpers.py]: If we get the `RequestedRangeNotSatisfiable` we are sure that the file will not be streamed, so at least in this case we can close it I think

Comment (untitaker) [inline flask/helpers.py]: My point is that we should do this regardless of whether we opened the file ourselves or got it passed as parameter.

Comment (RonnyPfannschmidt) [inline flask/helpers.py]: i wasn't aware that the sendfile thing would close the file automatically

i suppose in that case the file should be closed in any case, perhaps a note in the docs to ensure people are really aware

Comment (untitaker) [inline flask/helpers.py]: That's only what I'm inferring from the code. It does close the file in some cases and leave it open in the case where the file object is passed to `make_response`. However, `werkzeug.BaseResponse` appears to have its own mechanism for closing resources.

Comment (magne4000) [inline flask/helpers.py]: I had a warning in the test cases when I forgot to close the file with `RequestedRangeNotSatisfiable`

Comment (untitaker) [inline flask/helpers.py]: Okay, so I suspect it's expected that the file is automatically closed sooner or later. Not sure if there are actually any exceptions (other than the ones for incorrect invocations) that `make_response` raises.

**Comment by untitaker** [comment]
LGTM, anybody else want to take a look?

**Comment by untitaker** [comment]
Thanks!

**Pull request merged**